syntax = "proto3";

/* General Communication to Exoboot_Wrapper class
build command: python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. exoboot_remote.proto
*/

// General Messages
message null {}

message receipt {
    bool received = 1;
}

// Exoboot Controller Messages
message quit {
    bool mybool = 1;
}
message pause {
    bool mybool = 1;
}
message log {
    bool mybool = 1;
}
message torques {
    float peak_torque_left = 1;
    float peak_torque_right = 2;
}
message req_log {
    string thread = 1;
    string field = 2;
}
message ret_val {
    float val = 1;
}


// Logging Messages
message beaver {}

message testmsg {
    string msg = 1;
}

message subject_info {
    float startstamp = 1;
    string subjectID = 2;
    string trial_type = 3;
    string trial_cond = 4;
    string description = 5;
    bool usebackup = 6;
}

// Vickrey Auction Messages
message result {
    float t = 1;
    float subject_bid = 2;
    bool user_win_flag = 3;
    float current_payout = 4;
    float total_winnings = 5;
}

message survey {
    float t = 1;
    float enjoyment = 2;
    float rpe = 3;
}

// VAS Messages
message vas_info {
    float btn_num = 1;
    float trial = 2;
    float pres = 3;
}

message slider {
    float pitime = 1;
    repeated float torques = 2;
    repeated float mvs = 3;
}

message presentation {
    float btn_option = 1;
    float trial = 2;
    float pres = 3;
    repeated float torques = 4;
    repeated float values = 5; 
}

// JND Messages
message comparison {
    float pres = 1;
    float prop = 2;
    float T_ref = 3;
    float T_comp = 4;
    float truth = 5;
    float answer = 6;
}

// PREF Messages
message preference {
    float pres = 1;
    float torque = 2;
}

// Logging Messages 
// TODO IMPLEMENT GENERAL DICT LOGGING
message header {
    repeated string names = 1;
}

message type {
    repeated string types = 1;
}

message datastream {
    repeated string data = 1;
}


// GRPC Service
service exoboot_over_network {
    // General Methods
    rpc testconnection (testmsg) returns (receipt) {};
    rpc get_subject_info(null) returns (subject_info);
    rpc chop (beaver) returns (receipt) {}; // LoggingServer kill message

    // Exoboot Controller Commands
    rpc set_quit (quit) returns (receipt) {}
    rpc set_pause (pause) returns (receipt) {}
    rpc set_log (log) returns (receipt) {}
    rpc set_torque (torques) returns (receipt) {}
    rpc getpack (req_log) returns (ret_val) {}

    // Vickrey Auction Specific
    rpc call (result) returns (receipt) {};
    rpc question (survey) returns (receipt) {};

    //VAS Specific
    rpc update_vas_info (vas_info) returns (receipt) {};
    rpc slider_update (slider) returns (receipt) {};
    rpc presentation_result (presentation) returns (receipt) {};

    //JND Specific
    rpc comparison_result (comparison) returns (receipt) {};

    // Pref Specific
    rpc pref_result (preference) returns (receipt) {};
}
